modin/experimental/engines/omnisci_on_ray/frame/partition_manager.py

```diff
@@ -47,17 +48,95 @@ def _compute_num_partitions(cls):
         return 1
 
     @classmethod
-    def from_arrow(cls, at, return_dims=False):
+    def from_pandas(cls, df, return_dims=False, unsupported_cols=None):
+        if df.empty:
+            return super().from_pandas(df, return_dims), []
+
+        if unsupported_cols is None:
+            at, unsupported_cols = cls._get_unsupported_cols(df)
+        else:
+            return super().from_pandas(df, return_dims), unsupported_cols
+
+        if len(unsupported_cols) > 0:
+            # Putting pandas frame into partitions instead of arrow table, because we know
+            # that all of operations with this frame will be default to pandas and don't want
+            # unnecessaries conversion pandas->arrow->pandas
+            return super().from_pandas(df, return_dims), unsupported_cols
+        else:
+            # Since we already have arrow table, putting it into partitions instead
+            # of pandas frame, to skip that phase when we will be putting our frame to OmniSci
+            return cls.from_arrow(at, return_dims, unsupported_cols)
+
+    @classmethod
+    def from_arrow(cls, at, return_dims=False, unsupported_cols=None):
         put_func = cls._partition_class.put_arrow
 
         parts = [[put_func(at)]]
+        if unsupported_cols is None:
+            _, unsupported_cols = cls._get_unsupported_cols(at)
 
         if not return_dims:
-            return np.array(parts)
+            result = np.array(parts)
         else:
             row_lengths = [at.num_rows]
             col_widths = [at.num_columns]
-            return np.array(parts), row_lengths, col_widths
+            result = np.array(parts), row_lengths, col_widths
+
+        return result, unsupported_cols
+
+    @classmethod
+    def _get_unsupported_cols(cls, obj):
+        """
+        Finds columns with unsupported dtypes by OmniSci, and returns list of it.
+
+        Parameters
+        ----------
+            obj : pandas.DataFrame or pyarrow.Table, object to inspect on
+                unsupported column types.
+
+        Returns
+        -------
+            Tuple of pyarrow.Table representation of `obj` (for future using) and
+            a list of unsupported columns.
+        """
+        if isinstance(obj, (pandas.Series, pandas.DataFrame)):
+
+            def fast_select_dtypes(df, dtype, nrows=None):
+                cols = [i for i, col in enumerate(df.dtypes.items()) if col[1] == dtype]
+                if nrows is None:
+                    nrows = len(df)
```

Comment:
This would cause out-of-bounds access. Actually, I don't see why we need this parameter, we are not going to call this function with different values. We can simply always access the first row.

Reply:
done